### ISSUE TITLE
future of snackstack

### ISSUE THREAD

**Issue by amonks** [theuprising/snackpack#10]
tbh I think this snackstack shit is getting too complicated.

I think create-react-app has an infinitely better api, but:

- it doesn't support css preprocessing
- it doesn't handle server stuff
- it doesn't handle deployment

**Comment by amonks** [comment]
vague plan:

the goal is still to have all the common build crap live outside project repos.

Ideally, we include a single `snackstack` module, which provides a cli for testing, deployment, building, config, and ???.

that module should depend on a bunch of smaller tools, for

- production bundling
- testing
- deployment
- doc generation
- config/env whatever

these smaller tools should probably provide node apis and also cli interfaces. our main `snackstack` cli will use the node apis.

## snackpack

There are two things snackpack provides which I think we should be pushed out of scope and implemented more broadly/simply: config injection, and babel.

right now I have a script which runs snackpack on our snackship build files, then those build files call snackpack again on the project, then they call another instance of snackship for deployment. obviously that's fucked.

I think the reasons for all the indirection are twofold: config injection and babel configuration.

We need config vars in our build scripts _before_ building a bundle. We want to be able to write build scripts in es7. Webpack is cool for making a production bundle, but it's way overkill for babelifying our build scripts.

also

- snackpack is designed around how webpack works, rather than how compilation configuration should work
- automerge thing is too complicated
- environments stuff is too complicated
- logging sucks
- should we _really_ support custom builders?
  - rn we're using one for config and one to compile crumbles. config should just pull from process.env at build time.
- babel and postcss settings change a lot between projects, but other stuff doesn't. 
  - I think that "bundling for production" and "enabling modern syntax" are different tasks.

## babel stuff

**options**

- babel-register in the main cli entry point script
  - does that work with import/export?
- node-babel
- ???

## config injection

the validation stuff we have now is cool, but the injection is way too complicated.

The only time we don't have process.env is in the browser, and I think we can safely count on always running browser code through webpack. create-react-app uses webpack to inject variables from build-time ENV without any configuration wherever it sees `process.env` in your code.

I think I favor that approach. Most of our code (server code, scripts, all node stuff), it works out of the box. without a preprocessor. validation can be a library when needed.

**consideration:**

the server is built by webpack _and_ runs node, so it has two possible sources of ENV data: buildtime and runtime. runtime env is important on the server because of heroku-set variables like `MONGODB_URL`.

the browser only has buildtime ENV.

build scripts only have runtime ENV.

**consideration:**

I think we have several types of configuration.

1. things like which mailer to use, which _should_ be checked into source control
2. things like api keys, which probably _shouldn't_ be checked into source control
3. things like aws deployment api keys, which maybe _shouldn't_ even be on every dev's machine
4. things like `MONGODB_URL`, which need to come from runtime ENV in production, and could be checked into source control
5. things like [_______], which need to come from runtime ENV in production, and probably _shouldn't_ be checked into source control

**option:**

- type 1 config lives in project /config
- the app still declares what config it needs
- types 2-5 config happen in an .env file, which may or may not be passed around on slack.
- build scripts use the .env file
- the browser is injected with the .env file
- the server is injected with the .env file, and includes logic such that if runtime env is set, it always overrides

ideally, we have a vault running somewhere. every dev has their own key for the vault, which they set in ~/.*rc or store in the cli jawn, and which authorizes them to download a .env file with the values they need (maybe deployment stuff, maybe not). that does provide a single point of failure across all projects, so it'd have to be really secure. we probably shouldn't write it ourselves.

## deployment

I still don't get why we aren't using gulp tbh, but if not using gulp is a requirement then I think a library (snackship) with a bunch of js task wrappers is a good option.

I don't like snackship's `ship` function: I think just exporting a bunch of functions and writing regular scripts that use them is a saner approach.

**Comment by rongierlach** [comment]
love the idea of a cli, and would like to think about modularizing our default configs into little snacks that can be independently required akin to webpack's own loaders.

**Comment by amonks** [comment]
good call. "use built-ins or keep your own in project repo" is not a good story.

**Comment by amonks** [comment]
I think design-wise, a good way to approach this is readme-first.

1. write the readme we'd like to have for the main snackstack cli program
2. based on that, write the readme for each modular component it uses
3. based on that, write the components

Here's a wiki page where we can work on 1 together:

https://github.com/theuprising/snackpack/wiki/snackstack-readme.md

**Comment by kilokeith** [comment]
Like TDD, but the T is text.

**Comment by amonks** [comment]
http://tom.preston-werner.com/2010/08/23/readme-driven-development.html

**Comment by amonks** [comment]
my thinking re: snackstack-cli vs snackstack-components:

components should be totally agnostic, but the snackstack-cli should be totally opinionated, and feed its opinions into the components, if that makes sense.

**Comment by kilokeith** [comment]
It's the nacho cheese that covers the chips.

**Comment by rongierlach** [comment]
adding snacks (our little webpack config evolvers) should be easily doable via the cli, save configurations and pass them to the cli. built your lil dream webpack setup.

**Comment by amonks** [comment]
@rongierlach I disagree.

_snackpack_ should make it easy to add your own builders

but the _cheesy snackstack cli_ should be correct-out-of-the-box, rather than configurable.

More configuration means more divergence between projects, means it's harder to make changes.

**Comment by rongierlach** [comment]
snackstack components are what then exactly?

**Comment by amonks** [comment]
the libraries (like snackpack, snackship, snacktest, whatever) that the snackstack cli wraps

**Comment by amonks** [comment]
s/bunch of smaller tools/
